launch/launch/frontend/parser.py

```diff
@@ -163,12 +215,10 @@ def load(
 
         try:
             filename = getattr(fileobj, 'name', '')
-            # file extension without leading '.'
-            extension = os.path.splitext(filename)[1][1:]
-
-            sorted_parsers = sorted(cls.frontend_parsers.items())
-            implementations = [v for k, v in sorted_parsers if k == extension] + [
-                v for k, v in sorted_parsers if k != extension
+            implementations = cls.get_parsers_from_filename(filename)
+            implementations += [
+                parser for parser in cls.frontend_parsers.values()
+                if parser not in implementations
```

Comment:
So `implementations` is a list with the same items as `cls.frontend_parsers.values()`, but reordered so the first ones are the ones that support that extension, right?

I think that before we were doing the same (trying with all the parsers), so it's fine, but maybe if the extension is exactly known (e.g. yaml), we could only try that one and fail.

Reply:
> So `implementations` is a list with the same items as `cls.frontend_parsers.values()`, but reordered so the first ones are the ones that support that extension, right?
> 
> I think that before we were doing the same (trying with all the parsers), so it's fine,

yeah

> but maybe if the extension is exactly known (e.g. yaml), we could only try that one and fail.

I can create a follow-up issue if you want.

Reply:
> maybe if the extension is exactly known (e.g. yaml), we could only try that one and fail.

I lightly disagree. Extensions are non-binding IMHO.

Reply:
> I lightly disagree. Extensions are non-binding IMHO.

I don't understand what you mean.
There could be multiple extensions supporting a `.yaml` extension, and in that case we should try loading the file with them all.
What I mean is that it doesn't make sense to try to load a `.yaml` file with the XML frontend, I don't think anybody is going to name a xml file `my_launch_file.yaml`.


If we go ahead with this more strict approch, there are some extensions all plugins should support: no extension, .launch.
I agree this is unrelated to the PR though

Reply:
> What I mean is that it doesn't make sense to try to load a `.yaml` file with the XML frontend, I don't think anybody is going to name a `xml` file `my_launch_file.yaml`.

I think that's a reasonable enough convention to warrant the heuristic we have, but I wouldn't put restrictions we don't need. If e.g. a downstream user needs to annotate their launch XMLs in any way and use a different extension for whatever reason, I don't see why the XML frontend should be blocked from parsing it. And while it does make sense to be strict for some very ubiquitous extensions like `.yaml` or `.xml` (so common that using them for anything else is misleading), I don't love special rules.